### PR TITLE
Multiple orders can't be tracked properly when using different carriers

### LINKED ISSUE #22836: Multiple orders can't be tracked properly when using different carriers
#### Describe the bug

There are two product which has different shipping methods now if a guest place the order with these two products then guest is not able to follow both orders (orderreference#1  and orderreference#2) through the guest tracking page always getting open orderreference#1

#### Expected behavior

If the guest enter orderreference#1 then this order (# 1) details should visible and if the guest enter orderreference#2 then the order details should visible according  to orderreference#2

#### Steps to Reproduce

1. Set two standard product with different shipping methods
2. add these two products to the cart
3. Now complete the order as a guest
4. After completing you will get emails for different orders with the same reference
5. Now click on guest tracking and enter here order reference with # 1 or # 2 always you will get the order details that belongs to # 2, you will not able to check the order details for # 1 

**Screenshots**
Let's have an example
We have ordered "test product 1" and "test product 2" with the different shipping method as a guest and got two emails
order1

and 
order2

Now when we enter VKEMMGUQG#1 in order reference then get the order details with test product 2 

order1product

And when we enter VKEMMGUQG#2 

guesttracking

then still getting the same order details with test product 2 instead os test product 1
#### Additional information

* PrestaShop version: 1.7.7.1
* PHP version: 7.3

### PR DESCRIPTION
**Pull request by PeeyushAgrawal**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop 
| Description?      |fix issue - #22836
| Type?             | bug fix / major
| Category?         | FO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fixes #22836.
| Possible impacts? | guest order tracking.
| How to test?      | By adding these changes, follow the below steps.  -  
1. Set two standard product with different shipping methods
2. add these two products to the cart
3. Now complete the order as a guest
4. After completing you will get emails for different orders with the same reference
5. Now click on guest tracking and enter here order reference with # 1 or # 2 now you will able to get the order details according to # 1 or # 2

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/24195)

### CODE REVIEW

File: classes/order/Order.php [hunk 1]
```diff
@@ -1607,6 +1607,21 @@ public static function getByReferenceAndEmail($reference, $email)
 
         return new Order($id);
     }
+    
+    public static function getByReferenceEmailAndNumber($reference, $email, $ordernumber)
```
Comment (arouiadib) [inline classes/order/Order.php:1611]: I am not sure about the naming : ordernumber. Following arguments naming model in this same file, I think `$order_number` is better

File: classes/order/Order.php [hunk 2]
```diff
@@ -1607,6 +1607,21 @@ public static function getByReferenceAndEmail($reference, $email)
...
+    {
+        $sql = '
+          SELECT id_order
+            FROM `' . _DB_PREFIX_ . 'orders` o
+            LEFT JOIN `' . _DB_PREFIX_ . 'customer` c ON (o.`id_customer` = c.`id_customer`)
+                WHERE o.`reference` = \'' . pSQL($reference) . '\' AND c.`email` = \'' . pSQL($email) . '\'
+        ';
+        
+        $orderArray = Db::getInstance()->executeS($sql);
```
Comment (arouiadib) [inline classes/order/Order.php:1620]: maybe $orderList :thinking:

File: controllers/front/GuestTrackingController.php [hunk 3]
```diff
@@ -53,7 +53,12 @@ public function init()
      */
     public function postProcess()
     {
-        $order_reference = current(explode('#', Tools::getValue('order_reference')));
+        $orderRefArray = explode('#', Tools::getValue('order_reference'));
```
Comment (arouiadib) [inline controllers/front/GuestTrackingController.php:56]: same remarks as above about naming :thinking:

File: controllers/front/GuestTrackingController.php [hunk 4]
```diff
@@ -53,7 +53,12 @@ public function init()
...
+        $order_reference = current($orderRefArray);
+        $orderRefNumber = end($orderRefArray);
```
Comment (arouiadib) [inline controllers/front/GuestTrackingController.php:58]: Same

File: classes/order/Order.php (hunk 1 above)
Comment (arouiadib) [reply classes/order/Order.php:1611]: Perhaps the same SQL of the previous method can be  reused somehow avoiding code duplicate.

File: classes/order/Order.php [hunk 5]
```diff
@@ -1607,6 +1607,21 @@ public static function getByReferenceAndEmail($reference, $email)
 
         return new Order($id);
     }
+    
+    public static function getByReferenceEmailAndNumber($reference, $email, $ordernumber)
+    {
+        $sql = '
```
Comment (arouiadib) [inline classes/order/Order.php:1613]: Perhaps the same SQL of the previous method can be reused somehow avoiding code duplicate.

**Review by sowbiba — COMMENTED** [review]
Reading the code and the issue, the order_reference parameter will be something like `ORDERREF#1` (`ORDERREF#2`, `ORDERREF#3`, ...) without space
You're comparing the first and last item to know if there is only one order (if the `#1`, `#2` exists). If there are multiple orders, you make a query to get the order ids associated to the reference and return the good order based on his position

Am I right ?

I would add an ORDER BY to the query to be sure the orders are returned in the expected position

And please add comments to the code to explain what it's done. it can help later ;)

**Comment by sowbiba** [comment]
Please run `./vendor/bin/php-cs-fixer fix`

**Comment by PeeyushAgrawal** [comment]
> Reading the code and the issue, the order_reference parameter will be something like `ORDERREF#1` (`ORDERREF#2`, `ORDERREF#3`, ...) without space
> You're comparing the first and last item to know if there is only one order (if the `#1`, `#2` exists). If there are multiple orders, you make a query to get the order ids associated to the reference and return the good order based on his position
> 
> Am I right ?
> 
> I would add an ORDER BY to the query to be sure the orders are returned in the expected position
> 
> And please add comments to the code to explain what it's done. it can help later ;)

In  getByReferenceEmailAndNumber($reference, $email, $ordernumber)
the first parameter is reference, the second is email and the third is number.
For instance, suppose a guest has created order with three shipping methods so three orders will create.
Let order reference is ORDERREF then three orders will create with ORDERREF#1, ORDERREF#2, and ORDERREF#3

Now when guest will go to the guest tracking page and fill the form https://prnt.sc/12cy5o7 and if the guest enters ORDERREF#3 as a reference and guest@order.com as email then the value will be like - $reference = ORDERREF, $email = guest@order.com, $ordernumber = 3

File: classes/order/Order.php (hunk 1 above)
Comment (PierreRambaud) [reply classes/order/Order.php:1611]: You have to type hint the method.

**Comment by PierreRambaud** [comment]
> > Reading the code and the issue, the order_reference parameter will be something like `ORDERREF#1` (`ORDERREF#2`, `ORDERREF#3`, ...) without space
> > You're comparing the first and last item to know if there is only one order (if the `#1`, `#2` exists). If there are multiple orders, you make a query to get the order ids associated to the reference and return the good order based on his position
> > Am I right ?
> > I would add an ORDER BY to the query to be sure the orders are returned in the expected position
> > And please add comments to the code to explain what it's done. it can help later ;)
> 
> In getByReferenceEmailAndNumber($reference, $email, $ordernumber) the first parameter is reference, the second is email and the third is number. For instance, suppose a guest has created order with three shipping methods so three orders will create. Let order reference is ORDERREF then three orders will create with ORDERREF#1, ORDERREF#2, and ORDERREF#3
> 
> Now when guest will go to the guest tracking page and fill the form https://prnt.sc/12cy5o7 and if the guest enters ORDERREF#3 as a reference and [guest@order.com](mailto:guest@order.com) as email then the value will be like - $reference = ORDERREF, $email = [guest@order.com](mailto:guest@order.com), $ordernumber = 3

Hi @PeeyushAgrawalWebkul 
Could you please apply feedback and add comments, I'm sorry but I don't understand what you're trying to do :thinking:

**Comment by PierreRambaud** [comment]
Hi @PeeyushAgrawalWebkul ,

Since we had no news from you for more than 30 days, I'll close this pull request. Feel free to reopen or open another one if you think it's still relevant.

Thanks!

**Comment by PeeyushAgrawal** [comment]
@PierreRambaud Simply I'm trying to solve the issue for guest order tracking.

If a guest creates an order with 2 different shipping method then right now guest can't track the order.

**Pull request closed without merging**